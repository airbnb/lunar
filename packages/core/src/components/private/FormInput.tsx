import React from 'react';
import { Omit } from 'utility-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';
import buildInputStyles from '../../themes/buildInputStyles';

export type IgnoreAttributes =
  | 'children'
  | 'dangerouslySetInnerHTML'
  | 'onChange'
  | 'value'
  // RDFa attributes
  | 'about'
  | 'datatype'
  | 'inlist'
  | 'prefix'
  | 'property'
  | 'resource'
  | 'typeof'
  | 'vocab'
  // Non-standard attributes
  | 'inputMode'
  | 'is'
  | 'radioGroup'
  | 'color'
  | 'results'
  | 'security';

export type Props<T = any> = {
  /** Decrease font size and padding. */
  compact?: boolean;
  /** Mark the field as important. */
  important?: boolean;
  /** Mark the field as invalid. */
  invalid?: boolean;
  /** Add "notranslate" className to prevent Google Chrome translation. */
  noTranslate?: boolean;
  /** Mark the field as optional. */
  optional?: boolean;
  /** Current value. */
  value?: string;
  /** Callback to access the underlying input DOM element. */
  wrappedRef?: React.Ref<T>;
};

export type InputProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, IgnoreAttributes> &
  Props<HTMLInputElement>;

export type SelectProps = Omit<React.SelectHTMLAttributes<HTMLSelectElement>, IgnoreAttributes> &
  Props<HTMLSelectElement>;

export type TextAreaProps = Omit<
  React.TextareaHTMLAttributes<HTMLTextAreaElement>,
  IgnoreAttributes
> &
  Props<HTMLTextAreaElement>;

export type PrivateProps = Props &
  WithStylesProps & {
    // Support everything for convenience
    [key: string]: any;
    children?: React.ReactNode;
    hasPrefix?: boolean;
    hasSuffix?: boolean;
    tagName: 'input' | 'select' | 'textarea';
  };

class FormInput extends React.Component<PrivateProps> {
  static defaultProps = {
    children: null,
    compact: false,
    disabled: false,
    hasPrefix: false,
    hasSuffix: false,
    hidden: false,
    important: false,
    invalid: false,
    noTranslate: false,
    optional: false,
    value: '',
  };

  render() {
    const {
      children,
      compact,
      disabled,
      hasPrefix,
      hasSuffix,
      hidden,
      id,
      important,
      invalid,
      noTranslate,
      wrappedRef,
      optional,
      styles,
      tagName: Tag,
      ...restProps
    } = this.props;
    const isSelect = Tag === 'select';
    const props: any = {
      ...restProps,
      ...css(
        styles.input,
        important && styles.input_important,
        compact && styles.input_compact,
        invalid && styles.input_invalid,
        disabled && styles.input_disabled,
        hidden && styles.input_hidden,
        isSelect && styles.select,
        isSelect && compact && styles.select_compact,
        hasPrefix && styles.input_hasPrefix,
        hasSuffix && styles.input_hasSuffix,
      ),
      disabled,
      id,
      required: !optional,
    };

    // Only populate when invalid, otherwise it will break some CSS selectors
    if (invalid) {
      props['aria-invalid'] = true;
      props['aria-describedby'] = `${id}-error`;
    }

    // Cannot use chidren for input/textarea as they are void elements
    if (isSelect) {
      props.children = children;
    }

    // Add magical className to prevent Google Chrome translation
    if (noTranslate) {
      props.className += ' notranslate';
    }

    // @ts-ignore [ts] JSX element type 'Component' does not have any construct or call signatures. [2604]
    return <Tag {...props} ref={wrappedRef} data-gramm="false" data-enable-grammarly="false" />;
  }
}

export default withStyles(buildInputStyles)(FormInput);
