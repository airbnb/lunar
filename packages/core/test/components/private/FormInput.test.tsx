import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import FormInput from '../../../src/components/private/FormInput';

describe('<FormInput />', () => {
  const props = {
    id: 'test',
    name: 'test',
    onChange() {},
  };

  it('renders an `input` tag', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" {...props} />);

    expect(wrapper.is('input')).toBe(true);
  });

  it('renders a `textarea` tag', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="textarea" {...props} />);

    expect(wrapper.is('textarea')).toBe(true);
  });

  it('renders noTranslate className', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="textarea" {...props} noTranslate />);

    expect(wrapper.prop('className')).toContain('notranslate');
  });

  it('renders a `select` tag', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="select" {...props} />);

    expect(wrapper.is('select')).toBe(true);
  });

  it('renders compact', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" compact {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders select compact', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="select" compact {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders important', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" important {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders children in `select` tag as options', () => {
    const wrapper = shallowWithStyles(
      <FormInput tagName="select" {...props}>
        <option value="a">A</option>
        <option value="b">B</option>
        <option value="c">C</option>
      </FormInput>,
    );

    expect(wrapper.children()).toHaveLength(3);
    expect(
      wrapper
        .children()
        .at(0)
        .is('option'),
    ).toBe(true);
  });

  it('doesnt render children for `input` tag', () => {
    const wrapper = shallowWithStyles(
      <FormInput tagName="input" {...props}>
        Foo
      </FormInput>,
    );

    expect(wrapper.children()).toHaveLength(0);
  });

  it('doesnt render children for `textarea` tag', () => {
    const wrapper = shallowWithStyles(
      <FormInput tagName="textarea" {...props}>
        Foo
      </FormInput>,
    );

    expect(wrapper.children()).toHaveLength(0);
  });

  it('sets disabled', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" disabled {...props} />);

    expect(wrapper.prop('disabled')).toBe(true);
  });

  it('sets hidden', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" hidden {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('sets id', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" {...props} id="foo" />);

    expect(wrapper.prop('id')).toBe('foo');
  });

  it('sets name', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" disabled {...props} />);

    expect(wrapper.prop('name')).toBe('test');
  });

  it('sets value', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" value="Why?" {...props} />);

    expect(wrapper.prop('value')).toBe('Why?');
  });

  it('sets required/optional', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" {...props} />);

    expect(wrapper.prop('required')).toBe(true);

    wrapper.setProps({
      optional: true,
    });

    expect(wrapper.prop('required')).toBe(false);
  });

  it('sets aria attributes if invalid', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" {...props} />);

    expect(wrapper.prop('aria-invalid')).toBeUndefined();
    expect(wrapper.prop('aria-describedby')).toBeUndefined();

    wrapper.setProps({
      invalid: true,
    });

    expect(wrapper.prop('aria-invalid')).toBe(true);
    expect(wrapper.prop('aria-describedby')).toBe('test-error');
  });

  it('passes extra props to element', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" type="email" {...props} />);

    expect(wrapper.prop('type')).toBe('email');
  });

  it('supports prefixes', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" hasPrefix {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('supports suffixes', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" hasSuffix {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('sets maxLength', () => {
    const wrapper = shallowWithStyles(<FormInput tagName="input" maxLength={20} {...props} />);

    expect(wrapper.prop('maxLength')).toBe(20);
  });
});
