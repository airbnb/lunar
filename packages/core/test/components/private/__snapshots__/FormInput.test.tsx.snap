// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormInput /> renders compact 1`] = `
<input
  className="input_1st4d84-o_O-input_compact_19gkl4s"
  data-enable-grammarly="false"
  data-gramm="false"
  disabled={false}
  id="test"
  name="test"
  onChange={[Function]}
  required={true}
  value=""
/>
`;

exports[`<FormInput /> renders important 1`] = `
<input
  className="input_1st4d84-o_O-input_important_7iy6bw"
  data-enable-grammarly="false"
  data-gramm="false"
  disabled={false}
  id="test"
  name="test"
  onChange={[Function]}
  required={true}
  value=""
/>
`;

exports[`<FormInput /> renders select compact 1`] = `
<select
  className="input_1st4d84-o_O-input_compact_19gkl4s-o_O-select_173na0f-o_O-select_compact_1lll1pq"
  data-enable-grammarly="false"
  data-gramm="false"
  disabled={false}
  id="test"
  name="test"
  onChange={[Function]}
  required={true}
  value=""
/>
`;

exports[`<FormInput /> sets hidden 1`] = `
<input
  className="input_1st4d84-o_O-input_hidden_i17vo6"
  data-enable-grammarly="false"
  data-gramm="false"
  disabled={false}
  id="test"
  name="test"
  onChange={[Function]}
  required={true}
  value=""
/>
`;

exports[`<FormInput /> supports prefixes 1`] = `
<input
  className="input_1st4d84-o_O-input_hasPrefix_omi9gr"
  data-enable-grammarly="false"
  data-gramm="false"
  disabled={false}
  id="test"
  name="test"
  onChange={[Function]}
  required={true}
  value=""
/>
`;

exports[`<FormInput /> supports suffixes 1`] = `
<input
  className="input_1st4d84-o_O-input_hasSuffix_1mimzu3"
  data-enable-grammarly="false"
  data-gramm="false"
  disabled={false}
  id="test"
  name="test"
  onChange={[Function]}
  required={true}
  value=""
/>
`;
